File: backend/application/config/configuration_certifier.py
"""Configuration Certifier (Phase 14.3.6).

Production dependency graph analysis, runtime validation, diagnostics aggregation,
subsystem health audit, and production certification engine.
"""

from datetime import datetime, timezone
import logging
from threading import RLock
from typing import List, Optional, Tuple

from backend.application.config.configuration_source_manager import ConfigurationSourceManager
from backend.application.config.models import (
    ConfigurationCertificationResult,
    ConfigurationDiagnostics,
    ConfigurationHealth,
    ConfigurationRuntimeState,
    ConfigurationStatistics,
)

logger = logging.getLogger(__name__)


class ConfigurationCertifier:
    """Production certification engine for Configuration Runtime."""

    def __init__(self, source_manager: Optional[ConfigurationSourceManager] = None) -> None:
        """Initialize ConfigurationCertifier using Constructor Dependency Injection.

        Args:
            source_manager: Optional ConfigurationSourceManager instance.
        """
        self._lock = RLock()
        self._source_manager = source_manager or ConfigurationSourceManager()

    def certify(self) -> ConfigurationCertificationResult:
        """Execute complete production certification audit across all runtime subsystems.

        Returns:
            ConfigurationCertificationResult: Certification result report.
        """
        with self._lock:
            passed = 0
            failed = 0
            issues: List[str] = []

            # 1. Source registry check
            sources = self._source_manager.registry.list_sources()
            if len(sources) > 0:
                passed += 1
            else:
                failed += 1
                issues.append("Certification Failure: No configuration sources registered in SourceRegistry.")

            # 2. Source priority uniqueness check
            priorities = [s.priority for s in sources]
            if len(priorities) == len(set(priorities)):
                passed += 1
            else:
                failed += 1
                issues.append("Certification Warning: Duplicate source priorities detected in SourceRegistry.")

            # 3. Active profile validity check
            try:
                active_profile = self._source_manager.get_active_profile()
                if active_profile is not None and active_profile.profile_name:
                    passed += 1
                else:
                    failed += 1
                    issues.append("Certification Failure: Active configuration profile is invalid or missing.")
            except Exception as e:
                failed += 1
                issues.append(f"Certification Failure: Error retrieving active profile: {str(e)}")

            # 4. Schema manager & Resolver check
            try:
                resolve_res = self._source_manager.resolve_all()
                if len(resolve_res.errors) == 0:
                    passed += 1
                else:
                    failed += 1
                    issues.append(f"Certification Failure: {len(resolve_res.errors)} errors encountered during schema resolution.")
            except Exception as e:
                failed += 1
                issues.append(f"Certification Failure: Resolver exception during certification: {str(e)}")

            # 5. Validator check
            try:
                val_res = self._source_manager.validate()
                if val_res.is_valid:
                    passed += 1
                else:
                    failed += 1
                    for err in val_res.errors:
                        issues.append(f"Certification Validation Error [{err.key}]: {err.message}")
            except Exception as e:
                failed += 1
                issues.append(f"Certification Failure: Validator exception during certification: {str(e)}")

            # 6. Feature flag subsystem health check
            feat_health = self._source_manager.feature_manager.health()
            if feat_health.is_healthy:
                passed += 1
            else:
                failed += 1
                issues.extend([f"Feature Flag Subsystem Issue: {i}" for i in feat_health.issues])

            # 7. Secret manager health check
            sec_health = self._source_manager.secret_manager.health()
            if sec_health.is_healthy:
                passed += 1
            else:
                failed += 1
                issues.extend([f"Secret Subsystem Issue: {i}" for i in sec_health.issues])

            # 8. Overall aggregate health check
            agg_health = self._source_manager.health()
            if agg_health.is_healthy:
                passed += 1
            else:
                failed += 1
                issues.extend(agg_health.issues)

            total_checks = passed + failed
            avail_pct = (passed / total_checks * 100.0) if total_checks > 0 else 0.0
            is_certified = failed == 0

            logger.info("Configuration certifier completed: certified=%s (passed=%d, failed=%d, avail=%.1f%%).", is_certified, passed, failed, avail_pct)

            return ConfigurationCertificationResult(
                is_certified=is_certified,
                checks_passed=passed,
                checks_failed=failed,
                issues=tuple(issues),
                availability_percentage=avail_pct,
                certified_at=datetime.now(timezone.utc),
            )

    def validate_runtime(self) -> bool:
        """Validate configuration runtime readiness."""
        cert = self.certify()
        return cert.is_certified

    def diagnostics(self) -> ConfigurationDiagnostics:
        """Get aggregate configuration diagnostics snapshot."""
        with self._lock:
            return self._source_manager.diagnostics()

    def health(self) -> ConfigurationHealth:
        """Get aggregate health snapshot."""
        with self._lock:
            return self._source_manager.health()

    def statistics(self) -> ConfigurationStatistics:
        """Get aggregate statistics metrics snapshot."""
        with self._lock:
            return self._source_manager.statistics()

File: frontend/src/runtime/config/configuration_certifier.ts
/**
 * Configuration Certifier Engine (Phase 16.3.6).
 *
 * Implements end-to-end verification, dependency validation, source priority uniqueness audit,
 * performance benchmarking, score calculation, issue reporting, and production readiness certification
 * for the Frontend Configuration Runtime.
 */

import {
  CertificationHealth,
  CertificationIssue,
  CertificationReport,
  CertificationStatistics,
  ConfigurationCertification,
  ConfigurationRuntimeState,
  createCertificationHealth,
  createCertificationIssue,
  createCertificationReport,
  createCertificationStatistics,
  createConfigurationCertification,
  createConfigurationCertificationSummary,
} from './models';
import { IConfigurationProvider } from './interfaces';

export class ConfigurationCertifier {
  private readonly _provider: IConfigurationProvider;
  private _lastReport?: CertificationReport;

  private _certificationsRun = 0;
  private _passedCertifications = 0;
  private _failedCertifications = 0;
  private _totalScoreSum = 0;

  constructor(provider: IConfigurationProvider) {
    this._provider = provider;
  }

  public certify(): ConfigurationCertification {
    return this.runCertification().certification;
  }

  public runCertification(): CertificationReport {
    this._certificationsRun++;
    const issues: CertificationIssue[] = [];

    let totalChecks = 0;
    let passedChecks = 0;
    let failedChecks = 0;
    let warningChecks = 0;

    // 1. Runtime State Check
    totalChecks++;
    const state = this._provider.state();
    if (state.runtimeState === ConfigurationRuntimeState.READY) {
      passedChecks++;
    } else {
      failedChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'CRITICAL',
          component: 'RuntimeState',
          message: `Runtime is in state ${state.runtimeState}, expected READY.`,
          remediation: 'Initialize configuration provider before certification.',
        }),
      );
    }

    // 2. Source Registration & Priority Check
    totalChecks++;
    const sources = this._provider.listSources();
    const activeSources = sources.filter((s) => s.enabled);
    if (activeSources.length > 0) {
      passedChecks++;
    } else {
      warningChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'WARNING',
          component: 'Sources',
          message: 'No active configuration sources registered.',
          remediation: 'Register at least one active configuration source.',
        }),
      );
    }

    // Source Priority Uniqueness Check
    totalChecks++;
    const priorities = new Set<number>();
    let duplicatePriority = false;
    for (const src of sources) {
      if (priorities.has(src.priority)) {
        duplicatePriority = true;
        break;
      }
      priorities.add(src.priority);
    }
    if (!duplicatePriority) {
      passedChecks++;
    } else {
      warningChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'WARNING',
          component: 'Sources',
          message: 'Multiple sources share identical priority values.',
          remediation: 'Assign unique priority values to avoid nondeterministic source overrides.',
        }),
      );
    }

    // 3. Schema & Validation Engine Check
    totalChecks++;
    const validationRes = this._provider.validate();
    if (validationRes.valid) {
      passedChecks++;
    } else {
      failedChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'ERROR',
          component: 'Validator',
          message: `Configuration validation failed with ${validationRes.errors.length} errors.`,
          remediation: 'Fix invalid configuration values to satisfy registered schema constraints.',
        }),
      );
    }

    // 4. Resolution Engine Performance Benchmark Check
    totalChecks++;
    const resolveStart = performance ? performance.now() : Date.now();
    for (let i = 0; i < 100; i++) {
      this._provider.getAll();
    }
    const resolveEnd = performance ? performance.now() : Date.now();
    const benchmarkMs = Math.round((resolveEnd - resolveStart) * 100) / 100;

    if (benchmarkMs < 50) {
      passedChecks++;
    } else {
      warningChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'WARNING',
          component: 'ResolutionEngine',
          message: `Resolution benchmark elapsed ${benchmarkMs}ms, exceeding target threshold 50ms.`,
          remediation: 'Optimize source lookup chains and reduce redundant object allocations.',
        }),
      );
    }

    // 5. Active Profile Check
    totalChecks++;
    const activeProf = this._provider.getActiveProfile();
    if (activeProf && activeProf.active) {
      passedChecks++;
    } else {
      failedChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'ERROR',
          component: 'ProfileManager',
          message: 'No active profile currently selected.',
          remediation: 'Activate a valid configuration profile.',
        }),
      );
    }

    // 6. Feature Flag Manager Check
    totalChecks++;
    const fHealth = this._provider.featureHealth();
    if (fHealth.healthy) {
      passedChecks++;
    } else {
      warningChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'WARNING',
          component: 'FeatureFlagManager',
          message: 'Feature flag evaluation reported unhealthy state.',
        }),
      );
    }

    // 7. Sensitive Configuration Redaction Protection Check
    totalChecks++;
    const sHealth = this._provider.sensitiveHealth();
    const diag = this._provider.diagnostics();
    const rawExposedInDiag = JSON.stringify(diag).includes('RAW_SECRET_LEAK_CHECK_FAIL');

    if (sHealth.healthy && !rawExposedInDiag) {
      passedChecks++;
    } else {
      failedChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'CRITICAL',
          component: 'SecureConfigurationManager',
          message: 'Potential raw sensitive value exposure detected in diagnostics output.',
          remediation: 'Ensure all sensitive configuration properties are redacted before serialization.',
        }),
      );
    }

    // 8. Diagnostics Telemetry Aggregation Check
    totalChecks++;
    if (diag.health && diag.statistics && diag.capabilities) {
      passedChecks++;
    } else {
      failedChecks++;
      issues.push(
        createCertificationIssue({
          severity: 'ERROR',
          component: 'Diagnostics',
          message: 'Incomplete diagnostics telemetry reported.',
        }),
      );
    }

    // Compute score out of 100
    const rawScore = Math.max(0, Math.round((passedChecks / totalChecks) * 100));
    const criticals = issues.filter((i) => i.severity === 'CRITICAL').length;
    const errors = issues.filter((i) => i.severity === 'ERROR').length;
    const score = criticals > 0 ? Math.min(rawScore, 50) : errors > 0 ? Math.min(rawScore, 75) : rawScore;

    const certified = score >= 80 && criticals === 0 && errors === 0;

    if (certified) {
      this._passedCertifications++;
    } else {
      this._failedCertifications++;
    }
    this._totalScoreSum += score;

    const certification = createConfigurationCertification({
      certified,
      score,
      environment: this._provider.context().environment,
      issues: Object.freeze(issues),
      certifiedAt: new Date().toISOString(),
    });

    const summary = createConfigurationCertificationSummary({
      certified,
      score,
      totalChecks,
      passedChecks,
      failedChecks,
      warningChecks,
    });

    const report = createCertificationReport({
      certification,
      summary,
      diagnostics: diag,
      benchmarkMs,
      generatedAt: new Date().toISOString(),
    });

    this._lastReport = report;
    return report;
  }

  public certificationReport(): CertificationReport | undefined {
    return this._lastReport;
  }

  public statistics(): CertificationStatistics {
    const avg =
      this._certificationsRun > 0 ? Math.round(this._totalScoreSum / this._certificationsRun) : 100;

    return createCertificationStatistics({
      certificationsRun: this._certificationsRun,
      passedCertifications: this._passedCertifications,
      failedCertifications: this._failedCertifications,
      averageScore: avg,
    });
  }

  public health(): CertificationHealth {
    const lastScore = this._lastReport ? this._lastReport.certification.score : 100;
    const healthy = lastScore >= 80;
    return createCertificationHealth({
      healthy,
      lastCertificationScore: lastScore,
      statusMessage: healthy
        ? 'Configuration runtime is certified and operational.'
        : `Certification failed with score ${lastScore}/100.`,
    });
  }
}
